File: Difficulty: Easy/Largest Zigzag Sequence/largest-zigzag-sequence.java
class Solution{
    int solve(int i, int j, int[][] mat, int[][] dp, int n, int m){
        if(i>=n)
            return 0;
        
        if(dp[i][j] != -1)
            return dp[i][j];
            
        int temp=0;
        for(int k=0; k<m; k++){
            if(k != j){
                temp=Math.max(temp, mat[i][j]+solve(i+1, k, mat, dp, n, m));
            }
        }
        
        dp[i][j]=temp;
        return dp[i][j];
        
    }
    
    public int zigzagSequence(int[][] mat){
        int n=mat.length;
        int m=mat[0].length;
        
        if(n == 1 && m == 1)
            return mat[0][0];
        
        
        int[][] dp=new int[n][m];
        for(int[] row: dp){
            Arrays.fill(row, -1);
        }
        
        int res=0;
        for(int j=0; j<m; j++){
            res=Math.max(res, solve(0, j, mat, dp, n, m));    
        }
        
        return res;
        
        
    }
}

File: Difficulty: Easy/Largest Zigzag Sequence/largest-zigzag-sequence.py
class Solution:
    def solve(self, i, j, mat, dp, n, m):
        if i>=n:
            return 0
        
        if dp[i][j] != -1:
            return dp[i][j]
            
        temp=0;
        for k in range(m):
            if k != j:
                temp=max(temp, mat[i][j]+self.solve(i+1, k, mat, dp, n, m))
            
        
        dp[i][j]=temp
        return dp[i][j]
    
    def zigzagSequence(self, mat):
        n=len(mat)
        m=len(mat[0])
        
        if n == 1 and m == 1:
            return mat[0][0]
        
        
        dp=[[-1]*m for _ in range(n)]
        
        res=0
        for j in range(m):
            res=max(res, self.solve(0, j, mat, dp, n, m))
        
        return res
        
        
        
        